File: backend/app/routers/embeddings.py
"""Embedding API routes — 去重检测和状态查询."""

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException
from pydantic import BaseModel
from sqlmodel import Session

from app.dependencies import get_db
from app.models.pack import Pack
from app.schemas.embedding import (
    DuplicateCheckRequest,
    DuplicateCheckResponse,
    EmbeddingStatsResponse,
    PackEmbeddingStatusResponse,
    ProcessEmbeddingResponse,
)
from app.services import embedding_service

router = APIRouter(prefix="/api/v1/embeddings", tags=["embeddings"])


class EmbeddingStatusResponse(BaseModel):
    processing_packs: list[dict]


@router.get("/status", response_model=EmbeddingStatusResponse)
def get_embedding_status(db: Session = Depends(get_db)):
    """查询正在处理的 Pack 状态."""
    processing = embedding_service.get_embedding_status(db)
    return EmbeddingStatusResponse(processing_packs=processing)


@router.get("/stats", response_model=EmbeddingStatsResponse)
def get_embedding_stats(db: Session = Depends(get_db)):
    """查询 embedding 覆盖统计."""
    stats = embedding_service.get_embedding_stats(db)
    return EmbeddingStatsResponse(**stats)


@router.get(
    "/pack/{pack_id}/status",
    response_model=PackEmbeddingStatusResponse,
)
def get_pack_embedding_status(pack_id: int, db: Session = Depends(get_db)):
    """查询指定 Pack 的 embedding 处理状态."""
    status = embedding_service.get_pack_embedding_status(db, pack_id)
    if not status:
        raise HTTPException(
            status_code=404,
            detail=f"Pack {pack_id} not found or has no images",
        )
    return PackEmbeddingStatusResponse(**status)


@router.post(
    "/process/{pack_id}",
    response_model=ProcessEmbeddingResponse,
)
def process_pack_embeddings(
    pack_id: int,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
):
    """触发指定 Pack 的 embedding 处理（异步）."""
    pack = db.get(Pack, pack_id)
    if not pack:
        raise HTTPException(status_code=404, detail=f"Pack {pack_id} not found")
    background_tasks.add_task(
        embedding_service.process_remaining_embeddings,
        db,
        pack_id,
    )
    return ProcessEmbeddingResponse(pack_id=pack_id, status="queued")


@router.post("/check-duplicate", response_model=DuplicateCheckResponse)
async def check_duplicate(
    body: DuplicateCheckRequest,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
):
    """检测 Pack 是否与现有 Pack 重复.

    - 抽样 5 张图片进行检测
    - 如果没有重复，在后台异步处理剩余图片的 embedding
    """
    duplicates = await embedding_service.check_pack_duplicates(db, body.pack_id)

    if not duplicates:
        background_tasks.add_task(
            embedding_service.process_remaining_embeddings,
            db,
            body.pack_id,
        )

    return DuplicateCheckResponse(
        pack_id=body.pack_id,
        has_duplicates=len(duplicates) > 0,
        duplicates=duplicates,
    )

File: frontend/src/lib/api/embeddings.ts
import { clientFetch, serverFetch } from "./client";
import type {
  DuplicateCheckResult,
  EmbeddingStatus,
  EmbeddingStats,
  PackEmbeddingStatus,
  ProcessEmbeddingResult,
} from "./types";

export async function checkDuplicate(
  packId: number,
  isServer: boolean = false
): Promise<DuplicateCheckResult> {
  const fetch = isServer ? serverFetch : clientFetch;
  return fetch("/api/v1/embeddings/check-duplicate", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ pack_id: packId }),
  });
}

export async function getEmbeddingStatus(
  isServer: boolean = false
): Promise<EmbeddingStatus> {
  const fetch = isServer ? serverFetch : clientFetch;
  return fetch("/api/v1/embeddings/status");
}

export async function getEmbeddingStats(
  isServer: boolean = false
): Promise<EmbeddingStats> {
  const fetch = isServer ? serverFetch : clientFetch;
  return fetch("/api/v1/embeddings/stats");
}

export async function getPackEmbeddingStatus(
  packId: number,
  isServer: boolean = false
): Promise<PackEmbeddingStatus> {
  const fetch = isServer ? serverFetch : clientFetch;
  return fetch(`/api/v1/embeddings/pack/${packId}/status`);
}

export async function triggerEmbedding(
  packId: number
): Promise<ProcessEmbeddingResult> {
  return clientFetch(`/api/v1/embeddings/process/${packId}`, {
    method: "POST",
  });
}
